numpy/core/shape_base.py

```diff
@@ -423,32 +423,81 @@ def format_index(index):
         return parent_index, _nx.ndim(arrays)
 
 
-def _block(arrays, max_depth, result_ndim):
+def _assignment_recursion(result, arrays, list_ndim, start_indicies,
+                          depth=0):
+    if depth < list_ndim:
+        concat_depth = depth + result.ndim - list_ndim
+        for arr in arrays:
+            shape = _assignment_recursion(result, arr, list_ndim,
+                                          start_indicies, depth=depth+1)
+            start_indicies[concat_depth] += shape[concat_depth]
+        # lists are mutable, so we have to reset the start_indicies list
+        start_indicies[concat_depth:] = (0, ) * (result.ndim - concat_depth)
+        return shape
```

Comment:
This looks off to me - doesn't it just return the shape of the very last array?

Reply:
Feel free to comment it out, then run pytest. 

It took me a while to figure out what was wrong. It is the same reason why you don't have a function signature with

```
def foo(baz, bar=[]):
    bar.append(baz)
```

I use a to have a tuple based implementation, but it looked worse IMO
https://github.com/numpy/numpy/pull/11971/commits/2a3bd7fc92617c6b48a6a7df5eb18c8f79b206e4